From: SeongJae Park <sj@kernel.org>
Date: Tue, 30 Dec 2025 16:56:39 -0800
Subject: [PATCH] Revert "mm/damon/core: remove call_control in inactive
 contexts"

This reverts commit bb8068702decb2ee3ddb41ec814c1cab6fde8b4f.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 31 +------------------------------
 1 file changed, 1 insertion(+), 30 deletions(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 65482a0ce20b..2d3e8006db50 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -1442,35 +1442,6 @@ bool damon_is_running(struct damon_ctx *ctx)
 	return running;
 }
 
-/*
- * damon_call_handle_inactive_ctx() - handle DAMON call request that added to
- *				      an inactive context.
- * @ctx:	The inactive DAMON context.
- * @control:	Control variable of the call request.
- *
- * This function is called in a case that @control is added to @ctx but @ctx is
- * not running (inactive).  See if @ctx handled @control or not, and cleanup
- * @control if it was not handled.
- *
- * Returns 0 if @control was handled by @ctx, negative error code otherwise.
- */
-static int damon_call_handle_inactive_ctx(
-		struct damon_ctx *ctx, struct damon_call_control *control)
-{
-	struct damon_call_control *c;
-
-	mutex_lock(&ctx->call_controls_lock);
-	list_for_each_entry(c, &ctx->call_controls, list) {
-		if (c == control) {
-			list_del(&control->list);
-			mutex_unlock(&ctx->call_controls_lock);
-			return -EINVAL;
-		}
-	}
-	mutex_unlock(&ctx->call_controls_lock);
-	return 0;
-}
-
 /**
  * damon_call() - Invoke a given function on DAMON worker thread (kdamond).
  * @ctx:	DAMON context to call the function for.
@@ -1501,7 +1472,7 @@ int damon_call(struct damon_ctx *ctx, struct damon_call_control *control)
 	list_add_tail(&control->list, &ctx->call_controls);
 	mutex_unlock(&ctx->call_controls_lock);
 	if (!damon_is_running(ctx))
-		return damon_call_handle_inactive_ctx(ctx, control);
+		return -EINVAL;
 	if (control->repeat)
 		return 0;
 	wait_for_completion(&control->completion);
-- 
2.47.3

